### PR TITLE
ngram: add class and test

### PR DESCRIPTION
**Pull request by spoilt333**

### CODE REVIEW

File: moses/baselines/ngram/ngram.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import numpy as np
+import moses
+from moses import CharVocab
+from tqdm import tqdm
+
+
+class Model:
```
Comment (danpol) [inline moses/baselines/ngram/ngram.py]: Let's call it something like NGram

File: moses/baselines/ngram/ngram.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        self.n_tokens = len(self.vocab)
+        self.l_smooth = l_smooth
+        self._dict = dict()
+        self.default_probs = np.hstack([np.ones(self.n_tokens-4),
+                                        np.array([0., 1., 0., 0.])]) * l_smooth
+        self.fit(dataset)
+    
+    def fit(self, data):
+        with tqdm(total=len(data)) as pbar:
+            for i, line in enumerate(data):
```
Comment (danpol) [inline moses/baselines/ngram/ngram.py]: You can use tqdm directly around iterables: `tqdm(enumerate(data), total=len(data))`

File: moses/baselines/ngram/ngram.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import numpy as np
+import moses
+from moses import CharVocab
+from tqdm import tqdm
```
Comment (danpol) [inline moses/baselines/ngram/ngram.py]: Use `from tqdm.auto import tqdm`. In this case, tqdm will automatically choose which version to use — Jupyter or console

File: moses/baselines/ngram/ngram.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                context = context[1:]
+            probs = self._dict[context]
+            normed = probs / probs.sum()
+            next_symbol = np.random.choice(self.n_tokens, p=normed)
+            res.append(next_symbol)
+        
+        return self.vocab.ids2string(res)
+    
+    def likelihood(self, smiles):
+        pass
```
Comment (danpol) [inline moses/baselines/ngram/ngram.py]: TODO, I guess

File: moses/baselines/ngram/ngram.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                    for i in range(len(t_line)):
+                        if i + shift + 1 >= len(t_line):
+                            break
+                        k = t_line[i:i+shift+1]
+                        v = t_line[i+shift+1]
+                        probs = self._dict.get(k, self.default_probs.copy())
+                        probs[v] += 1.
+                        self._dict[k] = probs
+    
+    def generate_one(self):
```
Comment (danpol) [inline moses/baselines/ngram/ngram.py]: add `max_len` parameter to ensure we finish at some point. Use `max_len=100` by default

File: moses/baselines/ngram/ngram.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        res = [self.vocab.c2i[self.vocab.ss.bos]]
```
Comment (danpol) [inline moses/baselines/ngram/ngram.py]: self.vocab.bos

File: moses/baselines/ngram/ngram.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+        while res[-1] != self.vocab.c2i[self.vocab.ss.eos]:
```
Comment (danpol) [inline moses/baselines/ngram/ngram.py]: self.vocab.eos

File: moses/baselines/ngram.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        self._dict = dict()
+        self.vocab = None
+        self.default_probs = None
+
+    def fit(self, data):
+        self.vocab = CharVocab.from_data(data)
+        self.default_probs = np.hstack([np.ones(len(self.vocab)-4),
+                                        np.array([0., 1., 0., 0.])])
+        self.zero_probs = np.zeros(len(self.vocab))
+        print('fitting...')
```
Comment (danpol) [inline moses/baselines/ngram.py]: Add verbose option

File: moses/baselines/ngram.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        for line in tqdm(data, total=len(data)):
+            t_line = tuple(self.vocab.string2ids(line, True, True))
+            for i in range(len(t_line)):
+                for shift in range(self.max_context_len):
+                    if i + shift + 1 >= len(t_line):
+                        break
+                    context = t_line[i:i+shift+1]
+                    cid = t_line[i+shift+1]
+                    probs = self._dict.get(context, self.zero_probs.copy())
+                    probs[cid] += 1.
```
Comment (danpol) [inline moses/baselines/ngram.py]: You can use just 1. It is Python 3 :)

File: moses/baselines/ngram.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
```
Comment (danpol) [inline moses/baselines/ngram.py]: 1

File: moses/baselines/ngram.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+                    self._dict[context] = probs
+    
+    def generate_one(self, l_smooth=1., context_len=None, max_len=100):
```
Comment (danpol) [inline moses/baselines/ngram.py]: 1

File: moses/baselines/ngram.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+                context = context[1:]
+            probs = self._dict[context] 
+            probs += self.default_probs*l_smooth
+            normed = probs / probs.sum()
+            next_symbol = np.random.choice(len(self.vocab), p=normed)
+            res.append(next_symbol)
+        
+        return self.vocab.ids2string(res)
+    
+    def nll(self, smiles, l_smooth=1., context_len=None):
```
Comment (danpol) [inline moses/baselines/ngram.py]: 1

File: moses/baselines/ngram.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        for i in range(1, len(tokens)):
+            begin_index = max(i-context_len, 0)
+            context = tokens[begin_index:i]
+            while context not in self._dict:
+                context = context[1:]
+            
+            probs = self._dict[context] + self.default_probs
+            normed = probs / probs.sum()
+            prob = normed[tokens[i]]
+            if prob == 0.:
```
Comment (danpol) [inline moses/baselines/ngram.py]: Float comparisons... Maybe somehow use None to indicate that there were no samples?

File: moses/baselines/ngram.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+def reproduce():
+    data = moses.get_dataset('train')
+    model = NGram()
+    model.fit(data)
+    
+    for seed in [1, 2, 3]:
+        np.random.seed(seed)
+        smiles = model.generate(30000)
+        with open('../../data/samples/n_gram/n_gram_%d.csv' % (seed+3), 'w') as out:
```
Comment (danpol) [inline moses/baselines/ngram.py]: Why seed+3?

File: moses/baselines/ngram.py [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+            likelihood -= np.log(prob)
+        
+        return likelihood
+    
+    def generate(self, n, l_smooth=1., context_len=None, max_len=100):
+        generator = (self.generate_one(l_smooth=1., context_len=None, max_len=100) for i in range(n))
+        print('generating...')
+        return list(tqdm(generator, total=n))
+
+def reproduce():
```
Comment (danpol) [inline moses/baselines/ngram.py]: Add main call (if __name__ == "__main__") with this call
